### DISCUSSION TITLE
Possible problem with lora radio drivers for STM32WL

### DISCUSSION THREAD

**Discussion by patrykkk1** [zephyrproject-rtos/zephyr#58385] [Q&A]
**Introduction**
Hi,
I am developing firmware for a device with LoRaWAN based communication on the STM32WL55CCU6 microcontroller. So far everything seemed fine, without any major problems until I've left the device working for several days straight. After long  operation for several days there were situation in which the device stopped sending periodic uplink messages, and also there was no reaction to the transmitted downlink messages. However, the device was not completely dead, it was "partially" working. What I mean by that is some threads seemed to be operational. For example the device status LED, which was implemented in a separate UI thread was blinking continuously.  After some debugging, I suspect that the problem is related to the radio driver, but as I am not sure, instead of creating Bug Issue I decided to put it here as discussion first. Perhaps it is due to some incorrect use of LoRaWAN API by my application. As for my application, it collects some data from sensors and then sends it from a separate thread (not from  System Work Queue). Downlink messages and the Join procedure are also performed in a separate threads. Uplink messages are sent periodically, most of them are sent without acknowledgment, and after 5 unacknowledged messages, 1 acknowledgment message is sent to make sure that they are correctly delivered to the server. In case of no response from the server or if no downlink message has been received for a long time, the Join Request in transmitted as a last resort to restore communication.

**Describe the (potential?) bug**
From what I've discovered so far, the device gets stuck in the `SX126xWaitOnBusy` function, in a loop where it's waiting for the `RFBUSY` flag to disappear. The function definition is in `drivers/lora/sx126x.c`. For STM32WL, `SX126xWaitOnBusy` calls`sx126x_is_busy`, which reads the `RFBUSYS` bit directly from `PWRSR2` register using the ST LL driver.
I've attached debugger to the device and left it running while waiting for the moment when this situation would occur again. After more than 2 days I managed to catch this with debug session active. Below is the backtrace from GDB (there might be some differences with the line numbers as I might have added logs in some places during the debug process):
Looking at the backtrace, the device got stuck while it was performing the Join procedure. This procedure was caused by problem with gateway we had in our office and as there was no acknowledges received by the device for a long time it issued a Join Request. In the active GDB session, I've also checked the fields of the `dev_data` structure and as you can see, the radio is in `MODE_SLEEP` mode, which may explain why the `RFBUSY` status does not change.
In the `sx126x_spi_transceive` function, before sending the SPI command, the `SX126xCheckDeviceReady` function is called, in which the radio operating mode is checked and if the radio is in `MODE_SLEEP` mode, it is woken up, and only then the RFBUSY flag is checked. After completing the transaction, if it was not actually a command for setting the radio to sleep mode (`req_tx[0] != RADIO_SET_SLEEP`), the RFBUSY is checked and this is where the device gets stuck. It seems to me that after the transaction is done, the radio is somehow put into `MODE_SLEEP` somewhere.
That's all I've got for the moment, I will certainly continue with this issue, but if someone has any suggestions how should I approach this problem or what it might be related to I would really appreciate any help.

**To reproduce**
For now I’m not really sure how to reproduce this as it happens rarely (sometimes after few days after continuous device operation) which makes it harder to debug or find a way to reproduce this on demand.

**Logs and console output**
GDB backtrace:
```
#0  SX126xWaitOnBusy () at /workdir/zephyr/drivers/lora/sx126x.c:372
#1  0x08014088 in sx126x_spi_transceive (
    req_tx=req_tx@entry=0x200073ac <Lorawan::Network::JoinThreadStack+1516> "\212?\001\b", 
    req_rx=req_rx@entry=0x0, req_len=req_len@entry=1, 
    data_tx=data_tx@entry=0x200073bf <Lorawan::Network::JoinThreadStack+1535>, data_rx=data_rx@entry=0x0, 
    data_len=1) at /workdir/zephyr/drivers/lora/sx126x.c:135
#2  0x080271c6 in SX126xWriteCommand (opcode=opcode@entry=RADIO_SET_PACKETTYPE, 
    buffer=buffer@entry=0x200073bf <Lorawan::Network::JoinThreadStack+1535> "\001\344[", size=size@entry=1)
    at /workdir/zephyr/drivers/lora/sx126x.c:208
#3  0x0801452e in SX126xSetPacketType (packetType=<optimized out>, packetType@entry=PACKET_TYPE_LORA)
    at /workdir/modules/lib/loramac-node/src/radio/sx126x/sx126x.c:520
#4  0x0801495a in RadioSetModem (modem=<optimized out>)
    at /workdir/modules/lib/loramac-node/src/radio/sx126x/radio.c:570
#5  0x08014d06 in RadioSetTxConfig (modem=<optimized out>, power=<optimized out>, fdev=<optimized out>, 
    bandwidth=<optimized out>, datarate=12, coderate=1 '\001', preambleLen=8, fixLen=false, crcOn=true, 
    freqHopOn=false, hopPeriod=0 '\000', iqInverted=false, timeout=4000)
    at /workdir/modules/lib/loramac-node/src/radio/sx126x/radio.c:847
#6  0x0801a96a in RegionEU868TxConfig (
    txConfig=txConfig@entry=0x20007468 <Lorawan::Network::JoinThreadStack+1704>, 
    txPower=txPower@entry=0x20007467 <Lorawan::Network::JoinThreadStack+1703> "", 
    txTimeOnAir=txTimeOnAir@entry=0x20005844 <MacCtx+1048>)
    at /workdir/modules/lib/loramac-node/src/mac/region/RegionEU868.c:571
#7  0x08029024 in RegionTxConfig (region=<optimized out>, 
    txConfig=txConfig@entry=0x20007468 <Lorawan::Network::JoinThreadStack+1704>, 
    txPower=txPower@entry=0x20007467 <Lorawan::Network::JoinThreadStack+1703> "", 
    txTimeOnAir=txTimeOnAir@entry=0x20005844 <MacCtx+1048>)
    at /workdir/modules/lib/loramac-node/src/mac/region/Region.c:723
#8  0x0801714c in SendFrameOnChannel (channel=<optimized out>)
    at /workdir/modules/lib/loramac-node/src/mac/LoRaMac.c:3408
#9  0x080172e8 in ScheduleTx (allowDelayedTx=<optimized out>)
    at /workdir/modules/lib/loramac-node/src/mac/LoRaMac.c:3045
#10 0x080175b2 in SendReJoinReq (joinReqType=joinReqType@entry=JOIN_REQ)
    at /workdir/modules/lib/loramac-node/src/mac/LoRaMac.c:2849
#11 0x080190fc in LoRaMacMlmeRequest (
    mlmeRequest=mlmeRequest@entry=0x2000753c <Lorawan::Network::JoinThreadStack+1916>)
    at /workdir/modules/lib/loramac-node/src/mac/LoRaMac.c:5334
#12 0x080120c6 in lorawan_join_otaa (join_cfg=join_cfg@entry=0x200075b0 <Lorawan::Network::JoinThreadStack+2032>)
    at /workdir/zephyr/subsys/lorawan/lorawan.c:245
#13 0x080122ba in lorawan_join (join_cfg=0x200075b0 <Lorawan::Network::JoinThreadStack+2032>)
    at /workdir/zephyr/subsys/lorawan/lorawan.c:302
#14 0x080108e0 in Lorawan::Network::joinThreadRoutine ()
    at /workdir/application/app/src/lorawan/LorawanNetwork.cpp:232
#15 0x08025014 in z_thread_entry (entry=0x801087d <Lorawan::Network::joinThreadRoutine(void*, void*, void*)>, 
    p1=<optimized out>, p2=<optimized out>, p3=<optimized out>) at /workdir/zephyr/lib/os/thread_entry.c:36
#16 0x08025014 in z_thread_entry (entry=0x801087d <Lorawan::Network::joinThreadRoutine(void*, void*, void*)>, 
    p1=<optimized out>, p2=<optimized out>, p3=<optimized out>) at /workdir/zephyr/lib/os/thread_entry.c:36
```

`dev_data`:
```
$1 = {dio1_irq_callback = {node = {next = 0x0}, handler = 0x0, pin_mask = 0}, dio1_irq_work = {node = {
      next = 0x0}, handler = 0x8013e85 <sx126x_dio1_irq_work_handler>, queue = 0x20002420 <k_sys_work_q>, 
    flags = 0}, radio_dio_irq = 0x80148b1 <RadioOnDioIrq>, mode = MODE_SLEEP}
```

`SX126xWaitOnBusy` definiction:
```c
void SX126xWaitOnBusy(void)
{
    while (sx126x_is_busy(&dev_data)) {
        k_sleep(K_MSEC(1));
    } 
}
```

**Environment**
- Zephyr SDK 0.15.0 
- Zephyr kernel version 3.1.99 ([link](https://github.com/kuhnev/zephyr/tree/loramac-v4.7))

**Comment by 037vin** [comment]
Hi
Does anyone else think the WaitOnBusy function is a hole for a malfunctioning device to fall down?

**Comment by alinneacsu** [comment]
Any solution for this problem ?
